File: leetcode/47.全排列 II.go
func permuteUnique(nums []int) [][]int {
    ans := [][]int{}
    path := []int{}
    vis := make([]bool, len(nums))
    var dfs func()
    dfs = func() {
        if len(path) == len(nums) {
            ans = append(ans, append([]int(nil), path...))
            return
        }
        for i := 0; i < len(nums); i++ {
            if vis[i] {
                continue
            }
            if i > 0 && nums[i] == nums[i-1] && !vis[i -1] {
                continue
            }
            path = append(path, nums[i])
            vis[i] = true
            dfs()
            vis[i] = false
            path = path[: len(path) - 1]
        }
    }
    sort.Ints(nums)
    dfs()
    return ans
}

File: leetcode/47.全排列 II.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []
        t = []
        vis = [False] * n
        def dfs():
            if len(t) == n:
                ans.append(t[::])
                return
            for j in range(n):
                if vis[j] or j and nums[j] == nums[j-1] and not vis[j-1]:
                    continue
                t.append(nums[j])
                vis[j] = True
                dfs()
                t.pop()
                vis[j] = False
        nums.sort()
        dfs()
        return ans